### PR TITLE
7) added Services for Car and Driver

### PR DESCRIPTION
**Pull request by AntonPanchishko**

### CODE REVIEW

File: src/main/java/taxi/dao/CarDao.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+    List<Car> getAll();
+
+    Car update(Car car);
+
+    boolean delete(Long id);
+
+    void addDriverToCar(Driver driver, Car car);
+
+    void removeDriverFromCar(Driver driver, Car car);
```
Comment (Max-Pochepets) [inline src/main/java/taxi/dao/CarDao.java]: these methods shouldn't be in DAO layer

**Review by Max-Pochepets — CHANGES_REQUESTED** [review]

File: src/main/java/taxi/service/CarServiceImpl.java [hunk 2]
```diff
@@ -44,11 +47,24 @@ public void addDriverToCar(Driver driver, Car car) {
 
     @Override
     public void removeDriverFromCar(Driver driver, Car car) {
-        carDao.removeDriverFromCar(driver, car);
+        Storage.cars.stream()
```
Comment (Max-Pochepets) [inline src/main/java/taxi/service/CarServiceImpl.java]: you shouldn't access DB from Service. Use DAO

File: src/main/java/taxi/service/CarServiceImpl.java [hunk 3]
```diff
@@ -44,11 +47,24 @@ public void addDriverToCar(Driver driver, Car car) {
...
+                .removeIf(e -> Objects.equals(e, driver));
     }
 
     @Override
     public List<Car> getAllByDriver(Long driverId) {
-        return carDao.getAllByDriver(driverId);
+        List<Car> carsByDriver = new ArrayList<>();
+
+        for (int i = 0; i < Storage.cars.size(); i++) {
+            for (int j = 0; j < Storage.cars.get(j).getDrivers().size(); j++) {
```
Comment (Max-Pochepets) [inline src/main/java/taxi/service/CarServiceImpl.java]: same

**Review by Max-Pochepets — CHANGES_REQUESTED** [review]

**Review by Max-Pochepets — APPROVED** [review]

File: src/main/java/taxi/dao/CarDao.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package taxi.dao;
+
+import java.util.List;
+import taxi.model.Car;
+import taxi.model.Driver;
+
+public interface CarDao {
+    Car create(Car car);
+
+    Car get(Long id);
```
Comment (RomanDubovskyi) [inline src/main/java/taxi/dao/CarDao.java]: we return Optional on dao layer

File: src/main/java/taxi/dao/CarDao.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+    List<Car> getAll();
+
+    Car update(Car car);
+
+    boolean delete(Long id);
+
+    void addDriverToCar(Driver driver, Car car);
```
Comment (RomanDubovskyi) [inline src/main/java/taxi/dao/CarDao.java]: checklist

File: src/main/java/taxi/dao/CarDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+    @Override
+    public List<Car> getAll() {
+        return Storage.cars;
+    }
+
+    @Override
+    public Car update(Car car) {
+        Long currentCarId = car.getId();
+        Car oldCar = get(currentCarId);
```
Comment (RomanDubovskyi) [inline src/main/java/taxi/dao/CarDaoImpl.java]:
```suggestion
        Car oldCar = get(car.getId());
```

File: src/main/java/taxi/dao/DriverDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+                .findFirst().get();
+    }
+
+    @Override
+    public List<Driver> getAll() {
+        return Storage.drivers;
+    }
+
+    @Override
+    public Driver update(Driver driver) {
```
Comment (RomanDubovskyi) [inline src/main/java/taxi/dao/DriverDaoImpl.java:31]: same here

File: src/main/java/taxi/service/CarServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+        for (int i = 0; i < carDao.getAll().size(); i++) {
+            for (int j = 0; j < carDao.getAll().get(j).getDrivers().size(); j++) {
+                if (Objects.equals(carDao.getAll().get(j)
+                        .getDrivers().get(j).getId(), driverId)) {
+                    carsByDriver.add(Storage.cars.get(i));
+                }
+            }
+        }
+        return carsByDriver;
```
Comment (RomanDubovskyi) [inline src/main/java/taxi/service/CarServiceImpl.java]: let's move this logic to the dao layer, we should receive already filtered data from dao

File: src/main/java/taxi/service/CarServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    public void addDriverToCar(Driver driver, Car car) {
+        carDao.addDriverToCar(driver, car);
+    }
+
+    public void removeDriverFromCar(Driver driver, Car car) {
+        carDao.getAll().stream()
+                .filter(e -> Objects.equals(e, car))
+                .findFirst().get()
+                .getDrivers()
+                .removeIf(e -> Objects.equals(e, driver));
```
Comment (RomanDubovskyi) [inline src/main/java/taxi/service/CarServiceImpl.java]:
```suggestion
        car.getDrivers().remove(driver);
        carDao.update(car);
```

**Review by RomanDubovskyi — CHANGES_REQUESTED** [review]

File: src/main/java/taxi/dao/CarDao.java (hunk 5 above)
Comment (RomanDubovskyi) [reply src/main/java/taxi/dao/CarDao.java]: what about this comment?

**Review by RomanDubovskyi — CHANGES_REQUESTED** [review]

File: src/main/java/taxi/dao/CarDaoImpl.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    public List<Car> getAll() {
+        return Storage.cars;
+    }
+
+    @Override
+    public Car update(Car car) {
+        Long currentCarId = car.getId();
+        Car oldCar = get(car.getId()).get();
+        Storage.cars.set(Storage.cars.indexOf(oldCar), car);
+        return get(currentCarId).get();
```
Comment (RomanDubovskyi) [inline src/main/java/taxi/dao/CarDaoImpl.java:36]:
```suggestion
        Car oldCar = get(car.getId()).get();
        Storage.cars.set(Storage.cars.indexOf(oldCar), car);
        return get(car.getId).get();
```

File: src/main/java/taxi/dao/DriverDaoImpl.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+        Long currentDriverId = driver.getId();
```
Comment (RomanDubovskyi) [inline src/main/java/taxi/dao/DriverDaoImpl.java:32]:
```suggestion
```

**Review by RomanDubovskyi — APPROVED** [review]
fix comments before merging this PR

**Pull request merged**